### PR TITLE
new resource: azurerm_sentinel_alert_rule_action

### LINKED ISSUE #8128: Support for logic app playbooks in azurerm_sentinel_alert_rule_scheduled
### Community Note

* Please vote on this issue by adding a 👍 [reaction](https://blog.github.com/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/) to the original issue to help the community and maintainers prioritize this request
* Please do not leave "+1" or "me too" comments, they generate extra noise for issue followers and do not help prioritize the request
* If you are interested in working on this issue or have submitted a pull request, please leave a comment

### Description

Currently, the azurerm_sentinel_alert_rule_scheduled does not support configuring playbooks such as Logic-apps. This results in not being able to make automated responses through code, which takes away a large part of the automation of Azure Sentinel analytic queries and alerting. Playbooks should be supported to fully automate managing Sentinel alerts.

### New or Affected Resource(s)

* azurerm_sentinel_alert_rule_scheduled

### Potential Terraform Configuration

```hcl
data "azurerm_log_analytics_workspace" "example" {
  name                = "log-analytics1"
  resource_group_name = "my-resource-group"
}

data "azurerm_logic_app_workflow" "example" {
  name                = "workflow1"
  resource_group_name = "my-resource-group"
}

resource "azurerm_sentinel_alert_rule_scheduled" "example" {
  name                       = "example-with-playbook"
  log_analytics_workspace_id = data.azurerm_log_analytics_workspace.example.workspace_id
  display_name               = "example-with-playbook"
  severity                   = "High"
  logic_app_workflow_id      = data.azurerm_logic_app_workflow.example.id
  query                      = <<QUERY
AzureActivity |
  where OperationName == "Create or Update Virtual Machine" or OperationName =="Create Deployment" |
  where ActivityStatus == "Succeeded" |
  make-series dcount(ResourceId) default=0 on EventSubmissionTimestamp in range(ago(7d), now(), 1d) by Caller
QUERY
}
```

### References

https://www.terraform.io/docs/providers/azurerm/r/sentinel_alert_rule_scheduled.html

* #0000

### PR DESCRIPTION
**Pull request by magodo**
This PR implement one of the alert rule resource required in terraform-providers/terraform-provider-azurerm#6448. This also fixes #8128.

## Test Result

```bash
💤 via 🦉 v1.14.2 make testacc TEST=./azurerm/internal/services/sentinel/tests TESTARGS="-run='TestAccAzureRMSentinelAlertRuleAction_'"
==> Checking that code complies with gofmt requirements...
==> Checking that Custom Timeouts are used...
TF_ACC=1 go test ./azurerm/internal/services/sentinel/tests -v -run='TestAccAzureRMSentinelAlertRuleAction_' -timeout 180m -ldflags="-X=github.com/terraform-providers/terraform-provider-azurerm/version.ProviderVersion=acc"
=== RUN   TestAccAzureRMSentinelAlertRuleAction_basic
=== PAUSE TestAccAzureRMSentinelAlertRuleAction_basic
=== RUN   TestAccAzureRMSentinelAlertRuleAction_requiresImport
=== PAUSE TestAccAzureRMSentinelAlertRuleAction_requiresImport
=== CONT  TestAccAzureRMSentinelAlertRuleAction_basic
=== CONT  TestAccAzureRMSentinelAlertRuleAction_requiresImport
--- PASS: TestAccAzureRMSentinelAlertRuleAction_basic (194.22s)
--- PASS: TestAccAzureRMSentinelAlertRuleAction_requiresImport (222.36s)
PASS
ok      github.com/terraform-providers/terraform-provider-azurerm/azurerm/internal/services/sentinel/tests      222.406s
```

### CODE REVIEW

File: azurerm/internal/services/sentinel/parse/sentinel_alert_rules.go
Comment (tombuildsstuff) [inline azurerm/internal/services/sentinel/parse/sentinel_alert_rules.go:67]: rather than doing this, can we parse this and key into `alertRules` and `actions`?

File: azurerm/internal/services/sentinel/parse/sentinel_alert_rules_test.go
Comment (tombuildsstuff) [inline azurerm/internal/services/sentinel/parse/sentinel_alert_rules_test.go:153]: we should check the inverse of this too, if we expected one but didn't get one

File: azurerm/internal/services/sentinel/resource_arm_sentinel_alert_rule_action.go
Comment (tombuildsstuff) [inline azurerm/internal/services/sentinel/resource_arm_sentinel_alert_rule_action.go:54]: it might be worth making this `logic_app_workflow_id`?

File: azurerm/internal/services/sentinel/resource_arm_sentinel_alert_rule_action.go
Comment (tombuildsstuff) [inline azurerm/internal/services/sentinel/resource_arm_sentinel_alert_rule_action.go:98]: is it likely these could change in time/should be passed in?

File: azurerm/internal/services/sentinel/tests/resource_arm_sentinel_alert_rule_action_test.go
Comment (tombuildsstuff) [inline azurerm/internal/services/sentinel/tests/resource_arm_sentinel_alert_rule_action_test.go:170]: if this is a constant - is it worth making a new logic app trigger for `azurerm_logic_app_trigger_sentinel_alert`?

File: azurerm/internal/services/sentinel/parse/sentinel_alert_rules.go
Comment (magodo) [reply azurerm/internal/services/sentinel/parse/sentinel_alert_rules.go:67]: @tombuildsstuff I'm sorry I don't quite get your point here. Do you mean to embed `SentinelAlertRuleId` as a member inside `SentinelAlertRuleActionId`, and also embed `LogAnalyticsWorkspaceID` as a member inside `SentinelAlertRuleId`?

File: azurerm/internal/services/sentinel/parse/sentinel_alert_rules_test.go
Comment (magodo) [reply azurerm/internal/services/sentinel/parse/sentinel_alert_rules_test.go:153]: Oh, I missed that condition! THank you!

File: azurerm/internal/services/sentinel/resource_arm_sentinel_alert_rule_action.go
Comment (magodo) [reply azurerm/internal/services/sentinel/resource_arm_sentinel_alert_rule_action.go:54]: Yes, that was what I intended to do. But later I noticed in `azurerm_logic_app_trigger_xxx` (e.g. [this](https://www.terraform.io/docs/providers/azurerm/r/logic_app_trigger_custom.html#logic_app_id)) the same field is named after `logic_app_id`, so I follow that convention to be consistent.

File: azurerm/internal/services/sentinel/resource_arm_sentinel_alert_rule_action.go
Comment (magodo) [reply azurerm/internal/services/sentinel/resource_arm_sentinel_alert_rule_action.go:98]: This is exactly what the portal does. I feel user do not necessary need to care about the callback URL here.
On the other hand, if we want to pass it in, then do you think it make sense to expose a `Computed` field in `azurerm_logic_app_trigger_custom` and refer to it here?

Comment (magodo) [reply azurerm/internal/services/sentinel/resource_arm_sentinel_alert_rule_action.go:98]: Another way is that we specify anther field `logic_app_trigger_name`, meanwhile submit an issue to azure api spec to ask them make the API idempotent.

**Comment by magodo** [comment]
@tombuildsstuff Thank you for the review!
I have updated the code as I commented [above](https://github.com/terraform-providers/terraform-provider-azurerm/pull/6820#discussion_r423453483). Does this make sense?

With this change, it is possible to implement `azurerm_logic_app_trigger_sentinel_alert` (as we can give an customized `name` for this resource). While I'm not sure if that's a good idea from the maintenance POV. If we hard code the body in the resource implementation, then once the service changed that, we have to modify code. On the other hand, we only need to update the document and test to reflect the latest change.

What do you think?

**Comment by magodo** [comment]
Hi @tombuildsstuff would you please go on reviewing this PR and see if there is anything I need to do to get it merged?

**Comment by kaovd** [comment]
Any ETA on when this is getting merged? Last activity on this thread was two weeks ago.

**Comment by q-uu-rt** [comment]
This looks brilliant! Can't wait for it to come downstream so I don't have to use the az cli to implement this, will intergrate into my workflow nicely :)

File: azurerm/internal/services/sentinel/resource_arm_sentinel_alert_rule_action.go
Comment (magodo) [reply azurerm/internal/services/sentinel/resource_arm_sentinel_alert_rule_action.go:106]: Azure/azure-rest-api-specs#9424 is opened for this.

File: azurerm/internal/services/sentinel/parse/sentinel_alert_rules.go
Comment (tombuildsstuff) [inline azurerm/internal/services/sentinel/parse/sentinel_alert_rules.go:52]: can we switch this over to using an ID Formatter as per the other resources?

File: azurerm/internal/services/sentinel/parse/sentinel_alert_rules.go
Comment (tombuildsstuff) [inline azurerm/internal/services/sentinel/parse/sentinel_alert_rules.go:44]: can we remove this and use the ID Formatter interface instead for now?

File: azurerm/internal/services/sentinel/parse/sentinel_alert_rules.go
Comment (tombuildsstuff) [reply azurerm/internal/services/sentinel/parse/sentinel_alert_rules.go:67]: sorry that was unclear - as in, we should be using the same ID Parsers as everything else here - to ensure another ID isn't being used inadvertantly - so can we make this:

```suggestion
	action := SentinelAlertRuleActionId{
		ResourceGroup: id.ResourceGroup,
	}
	if server.Name, err = id.PopSegment("actions"); err != nil {
		return nil, err
	}
	
	(etc)
```

File: azurerm/internal/services/sentinel/tests/resource_arm_sentinel_alert_rule_action_test.go
Comment (tombuildsstuff) [inline azurerm/internal/services/sentinel/tests/resource_arm_sentinel_alert_rule_action_test.go:30]: mind adding the TODO here too?

File: azurerm/internal/services/sentinel/tests/resource_arm_sentinel_alert_rule_action_test.go
Comment (tombuildsstuff) [inline azurerm/internal/services/sentinel/tests/resource_arm_sentinel_alert_rule_action_test.go:143]: can we make this dynamic here?

File: website/docs/r/sentinel_alert_rule_action.html.markdown
Comment (tombuildsstuff) [inline website/docs/r/sentinel_alert_rule_action.html.markdown:18]: can we remove the provider block here?

File: website/docs/r/sentinel_alert_rule_action.html.markdown
Comment (tombuildsstuff) [inline website/docs/r/sentinel_alert_rule_action.html.markdown:92]: should this be `sentinel_alert_rule_id`?

File: azurerm/internal/services/loganalytics/parse/log_analytics_workspace.go
Comment (tombuildsstuff) [inline azurerm/internal/services/loganalytics/parse/log_analytics_workspace.go:10]: can we make this `SubscriptionId` here?

File: azurerm/internal/services/loganalytics/parse/log_analytics_workspace_test.go
Comment (tombuildsstuff) [inline azurerm/internal/services/loganalytics/parse/log_analytics_workspace_test.go:44]: this is the default value for a UUID, best to make the test use all 1's?

File: azurerm/internal/services/logic/parse/logic_app_workflow.go
Comment (tombuildsstuff) [inline azurerm/internal/services/logic/parse/logic_app_workflow.go:10]: this should be `SubscriptionId` here

File: azurerm/internal/services/logic/parse/logic_app_workflow.go
Comment (tombuildsstuff) [inline azurerm/internal/services/logic/parse/logic_app_workflow.go:15]: can we make this comply with the ResourceID interface for now?

File: azurerm/internal/services/logic/parse/logic_app_workflow_test.go
Comment (tombuildsstuff) [inline azurerm/internal/services/logic/parse/logic_app_workflow_test.go:38]: can we update the test data to use a non-default value here? for now we can likely remove this in favour of the ID Formatters

File: azurerm/internal/services/sentinel/parse/sentinel_alert_rules.go
Comment (tombuildsstuff) [inline azurerm/internal/services/sentinel/parse/sentinel_alert_rules.go:17]: can we switch this over to using the ID Formatters here?

File: azurerm/internal/services/sentinel/parse/sentinel_alert_rules.go
Comment (tombuildsstuff) [inline azurerm/internal/services/sentinel/parse/sentinel_alert_rules.go:11]: can we remove this/make this Subscription ID here?

File: azurerm/internal/services/sentinel/parse/sentinel_alert_rules_test.go
Comment (tombuildsstuff) [inline azurerm/internal/services/sentinel/parse/sentinel_alert_rules_test.go:53]: (as above) can we remove this for now, since this should be threaded through by the interface?

File: azurerm/internal/services/sentinel/resource_arm_sentinel_alert_rule_action.go
Comment (tombuildsstuff) [inline azurerm/internal/services/sentinel/resource_arm_sentinel_alert_rule_action.go:90]: isn't threading the RP through here a swagger bug?

**Review by tombuildsstuff — CHANGES_REQUESTED** [review]
hey @magodo 

Apologies for the delayed re-review here - I've taken a look through and left some comments inline, but this is otherwise looking pretty good 👍 

Thanks!

File: azurerm/internal/services/loganalytics/parse/log_analytics_workspace.go
Comment (magodo) [reply azurerm/internal/services/loganalytics/parse/log_analytics_workspace.go:10]: Instead, I removed that field as it is not used anywhere and we already has implement the IDFormatter for this type.

File: azurerm/internal/services/sentinel/resource_arm_sentinel_alert_rule_action.go
Comment (magodo) [reply azurerm/internal/services/sentinel/resource_arm_sentinel_alert_rule_action.go:90]: I'm sorry to be dumb here, what do you mean is a swagger bug?

**Comment by magodo** [comment]
@tombuildsstuff I have updated the code per your comments, except the last comment which I'm not so sure about the intent.

Regarding the IDFormatter implementation, I further changed the related resource/data source to set the returned value of `ID()` instead of setting the "raw" ID returned from response.

File: azurerm/internal/services/sentinel/resource_arm_sentinel_alert_rule_action.go
Comment (katbyte) [inline azurerm/internal/services/sentinel/resource_arm_sentinel_alert_rule_action.go:67]: could we validate this?

File: azurerm/internal/services/sentinel/resource_arm_sentinel_alert_rule_action.go
Comment (katbyte) [inline azurerm/internal/services/sentinel/resource_arm_sentinel_alert_rule_action.go:171]: Until this bug is fixed i think this resource is blocked?

File: azurerm/internal/services/sentinel/tests/resource_arm_sentinel_alert_rule_action_test.go
Comment (katbyte) [inline azurerm/internal/services/sentinel/tests/resource_arm_sentinel_alert_rule_action_test.go:15]: We are missing a complete and update test here?

**Review by katbyte — CHANGES_REQUESTED** [review]
Thanks for the PR @magodo - looks good but i think this is blocked until the service team fixes the service?

File: azurerm/internal/services/sentinel/tests/resource_arm_sentinel_alert_rule_action_test.go
Comment (magodo) [reply azurerm/internal/services/sentinel/tests/resource_arm_sentinel_alert_rule_action_test.go:15]: There is no Update function for this resource, hence only basic test is available.

File: azurerm/internal/services/sentinel/resource_arm_sentinel_alert_rule_action.go
Comment (magodo) [reply azurerm/internal/services/sentinel/resource_arm_sentinel_alert_rule_action.go:67]: There is no source telling what pattern should a trigger name be (the `azurerm_logic_app_trigger_custom` resource hasn't specify a validation for its name either). So I'm going to simply use the `validation.StringIsNotEmpty` here.

File: azurerm/internal/services/sentinel/resource_arm_sentinel_alert_rule_action.go
Comment (magodo) [reply azurerm/internal/services/sentinel/resource_arm_sentinel_alert_rule_action.go:171]: Currently, users will have to use the lifecycle metadata to ignore the `logic_app_trigger_name`, also the import doesn't work. 
So we might had better to wait the API issue is fixed. However, on the other hand, the swagger issue has been hanging for a relative long time, and I don't think someone will pick it up soon.

**Comment by magodo** [comment]
Hi @katbyte thank you for your comments! I have replied them above. The main concern is that the API issue might not be addressed soon, so that means the PR will be blocked for a unpredicatable long time, which might blocks some users.

**Comment by katbyte** [comment]
As this is blocked on https://github.com/Azure/azure-rest-api-specs/issues/9424 i am going to close this until the issues is resolved. At that point we can re-poen and re-review this.

File: azurerm/internal/services/sentinel/resource_arm_sentinel_alert_rule_action.go
Comment (NVolcz) [reply azurerm/internal/services/sentinel/resource_arm_sentinel_alert_rule_action.go:172]: Do we really need the `azurerm_sentinel_alert_rule_action` resource to be so complex? All it really does is provide a link between a Logic app and a Sentinel alert rule, it should therefore be enough with two values.

This PR proposes:
```
resource "azurerm_sentinel_alert_rule_action" "example" {
  name                   = "example-alert-rule-action"
  sentinel_alert_rule_id = azurerm_sentinel_alert_rule_scheduled.example.id
  logic_app_id           = azurerm_logic_app_trigger_custom.example.logic_app_id
  logic_app_trigger_name = azurerm_logic_app_trigger_custom.example.name
  depends_on             = [azurerm_logic_app_trigger_custom.example]
}
```

I suggest we simplify it to:
```
resource "azurerm_sentinel_alert_rule_action" "example" {
  sentinel_alert_rule_id = azurerm_sentinel_alert_rule_scheduled.example.id

// One of the value below:
  logic_app_id           = azurerm_logic_app_trigger_custom.example.logic_app_id
  logic_app_trigger_name = azurerm_logic_app_trigger_custom.example.name
}
```

It also seems incorrect to set the `logic_app_trigger_name` to the value of `TriggerUrl`

**Comment by NVolcz** [comment]
I believe we do not need the [Azure/azure-rest-api-specs#9424 ](https://github.com/Azure/azure-rest-api-specs/issues/9424) to be resolved in order to unblock this. See my comment: https://github.com/terraform-providers/terraform-provider-azurerm/pull/6820/#discussion_r517978122

**Pull request closed without merging**
